[karthickkc] iP

karthickkc opened the pull request:
# This is a heading
> Talk is cheap, show me the code ~ Linus Torvalds
- This
- is
- a 
- list
1. A
2. number
3. list
This is my `main` method:
```ruby
public static void main(String[] args) {
        Bro bro = new Bro();
        bro.run();
    }
```
My current *Tasks* :+1::
- [x] Event
- [ ] Deadline
- [ ] Todo
Link to my project: https://github.com/karthickkc/ip

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
did level-0

Commit:
did Level-1

Commit:
did Level-2

Commit:
did Level-3, added Task class

Commit:
did Level-4, added 3 subclasses for Task

Commit:
Testing

Commit:
handling erroneous input

Commit:
did Level-6

Commit:
cleaned up code and did more error handling

Commit:
Add save feature

As of now data is lost once the bot stops running and tasks are not saved and are unretrievable

Therefore by saving these task in a local file and then loading in the data when bot is run again, allows information that was previously entered to be retrieved.

Commit:
add date and time feature for task

date and time should be added for tasks with deadline or event with start and end dates and times instead of just strings

made use of java capabilities and libraries to do this and also fixed some bugs with the saving of tasks even after chatbot is closed

Commit:
Add more OOP principles

Everything was being done in the Bro.java file, so created more classes to handle different things.

The Ui class that handles all the printing required.
Storage which handles saving tasks and then loading them back in when the bot is run again.
Parser which handles parsing of user inputs.
TaskList which handles adding, deleting, marking, unmarking of tasks

Commit:
Add package

Created package called bro, and moved all the classes under that package except for Bro.java

Commit:
Merge branch 'add-gradle-support' of https://github.com/karthickkc/ip

Commit:
Add JUnit Tests

Added tests to see if addEvent throws Exception is wrong input is given.
Added test to see if parseDate can successfully parseDate is correct input is given

Commit:
Edit build.gradle

Somes names in shadowJar were wrong and the mainClass was not set correctly

Commit:
Add JavaDocs and coding standard

Added JavaDocs for most methods and tweaked code to follow coding standard

Commit:
Add find capability

Allows users to find tasks that contain a specific keyword using find command followed by keyword e.g. find book

Commit:
Add files for GUI

Added necessary files for GUI, but the code from older files remains unchanged

Commit:
Finish GUI Setup

Finished GUI Setup, mainly by changing all the ui commands to return String instead of printing to the screen. Ensured that files are still saved to and loading in properly. Did some aesthetic changes and changes to javadocs

jarrellc1 commented on src/main/java/Bro.java:
```diff
@@ -99,8 +99,84 @@ public void run() {
+                        return tasks.findTasks(info);
+                        //break;
+                    case "todo":
+                        String todo = tasks.addTodo(info);
```
I like how you’ve separated action from info, but I’m wondering if info could be renamed to something more descriptive, like taskDetails or commandArgs, to better reflect what it contains.

jarrellc1 commented on src/main/java/Bro.java:
```diff
@@ -99,8 +99,84 @@ public void run() {
+                }
+            } catch (BroException be) {
+                return be.getMessage();
+            } catch (NumberFormatException nfe) {
```
I like how you've caught specific exceptions like BroException, NumberFormatException, and IndexOutOfBoundsException.

jarrellc1 commented on src/main/java/MainWindow.java:
```diff
@@ -28,9 +36,29 @@ public void initialize() {
+        ui = new Ui();
+        parser = new Parser(ui);
+        tasks = new TaskList(ui, parser);
+        storage = new Storage("./src/main/java/data.txt", tasks);
```
I noticed that the file path for data.txt is hardcoded in storage = new Storage("./src/main/java/data.txt", tasks);. Would it be helpful to extract this path into a constant or configuration file so that it can be easily modified without changing the code?

jarrellc1 reviewed (commented):
Overall, the code is well written, some improvements can be more descriptive names.
Good job.

zacharyxzz commented on src/main/java/MainWindow.java:
```diff
@@ -0,0 +1,78 @@
+        } catch (FileNotFoundException fnfe) {
+            System.out.println("Nothing to load");
+        } catch (BroException be) {
+            System.out.println(be.getMessage());
```
Perhaps error messages that are more descriptive could be used here?

zacharyxzz commented on src/main/java/Bro.java:
```diff
@@ -0,0 +1,182 @@
+     *
+     * The state of tasks is saved to a file after every change.
+     * Handles exceptions for file loading errors, command format errors, and task index out-of-bounds.
+     */
```
I like how this description provides a clear and comprehensive overview of the application's functionality.

zacharyxzz reviewed (commented):
Great job overall!
I like how comprehensive and clear the JavaDocs are. Just some nits to consider fixing.

Commit:
Add assertions to validate key assumptions

Added assert statements to document and validate critical assumptions:
  - Ensure that input strings for creation of tasks have the necessary information
  - Validate that the task number to mark, unmark or delete is one and above
  - Check that the task string to load in contains the data needed

These assertions will help catch potential bugs early by ensuring that the input values meet the expected criteria. If an assertion fails, an AssertionError will be thrown, making it easier to diagnose issues during development.

Commit:
Refactor code for better code quality and readability

Removed large chunks of unused code in the Bro.java file as the chatbot has switched from CLI to GUI

Edited the message displayed by BroException, if exception is thrown, to look better in the GUI

Made some minor changes and removed unused methods in Ui.java

This refactor improves robustness and code quality, and enhances code maintainability for future development.

karthickkc committed:
Merge pull request #1 from karthickkc/branch-A-Assertions

Add assertions to validate key assumptions

Commit:
Merge branch 'master' of https://github.com/karthickkc/ip

* 'master' of https://github.com/karthickkc/ip:
  Add assertions to validate key assumptions

Commit:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Add assertions to validate key assumptions

karthickkc committed:
Merge pull request #2 from karthickkc/branch-A-CodeQuality

Branch a code quality

Commit:
Merge branch 'master' of https://github.com/karthickkc/ip

* 'master' of https://github.com/karthickkc/ip:
  Refactor code for better code quality and readability

Commit:
Add functionality to detect duplicates

Checks if the most recently added task is already present
in the list of tasks.

If the task is found, a message is returned indicating that
the task may be a duplicate and prompts the user for
further action.

Commit:
Add functionality to detect duplicates

Checks if the most recently added task is already present in the list of tasks.

If the task is found, a message is returned indicating that the task may be a duplicate and prompts the user for further action.

karthickkc committed:
Edit boolean variables

Some boolean variables had wrong naming
convention, so corrected these variables

karthickkc committed:
Merge branch 'master' of https://github.com/karthickkc/ip

# Conflicts:
#	src/main/java/Bro.java

karthickkc committed:
Shorten method

getResponse method was too long so
shortened it by abstracting another method
and doing some other changes

karthickkc committed:
Shorten method further

Abstracted one more method to
further shorten the getResponse
method

karthickkc committed:
Add Ui.png

Added image representative of the
chatbot

karthickkc committed:
Add UserGuide

Added user guide for the chat bot

Did some bugfixes

karthickkc committed:
Update Storage class

Change the path of the data.txt file
to be relative rather than hard-
coded
